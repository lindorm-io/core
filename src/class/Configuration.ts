import { mergeObjectsWithBias } from "../util/merge-with-bias";
import { switchOnEnvironment } from "../util/switch-on-environment";

export interface IConfigurationBase {
  NODE_ENVIRONMENT: string;
}

export interface IConfigurationOptions {
  production: IConfigurationBase;
  staging: IConfigurationBase;
  development: IConfigurationBase;

  environment: IConfigurationBase;
  test: IConfigurationBase;
}

export abstract class Configuration<Interface extends IConfigurationBase> {
  private production: IConfigurationBase;
  private staging: IConfigurationBase;
  private development: IConfigurationBase;

  private environment: IConfigurationBase;
  private test: IConfigurationBase;

  protected constructor(options: IConfigurationOptions) {
    this.production = options.production;
    this.staging = options.staging;
    this.development = options.development;

    this.environment = options.environment;
    this.test = options.test;
  }

  public get(environment: string): IConfigurationBase {
    const data: unknown = mergeObjectsWithBias(
      this.environment,
      switchOnEnvironment(environment, {
        production: this.production,
        staging: this.staging,
        development: this.development,
        test: this.test,
      }),
    );
    return data as IConfigurationBase;
  }
}
