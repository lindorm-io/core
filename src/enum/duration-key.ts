export enum DurationKey {
  YEARS = "years",
  MONTHS = "months",
  DAYS = "days",
  HOURS = "hours",
  MINUTES = "minutes",
  SECONDS = "seconds",
}
