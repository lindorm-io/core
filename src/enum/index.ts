export * from "./duration-key";
export * from "./node-environment";
