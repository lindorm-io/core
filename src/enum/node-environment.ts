export enum NodeEnvironment {
  PRODUCTION = "production",
  STAGING = "staging",
  DEVELOPMENT = "development",
  TEST = "test",
}
