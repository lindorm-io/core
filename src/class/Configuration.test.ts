import { NodeEnvironment } from "../enum";
import { Configuration, IConfigurationBase, IConfigurationOptions } from "./Configuration";

interface IMockConfiguration extends IConfigurationBase {
  SERVER_PORT: number;
  HOST: string;
  PRIO: string;
}

class MockConfig extends Configuration<IMockConfiguration> {
  constructor(options: IConfigurationOptions) {
    super(options);
  }
}

describe("Configuration", () => {
  let config: MockConfig;

  let environmentConfig: IMockConfiguration;
  let productionConfig: IMockConfiguration;
  let stagingConfig: IMockConfiguration;
  let developmentConfig: IMockConfiguration;
  let testConfig: IMockConfiguration;

  beforeAll(() => {
    environmentConfig = {
      NODE_ENVIRONMENT: null,
      SERVER_PORT: null,
      HOST: null,
      PRIO: "recognized",
    };
    productionConfig = {
      NODE_ENVIRONMENT: NodeEnvironment.PRODUCTION,
      SERVER_PORT: 2345,
      HOST: "https://lindorm.io/",
      PRIO: "ignored",
    };
    stagingConfig = {
      NODE_ENVIRONMENT: NodeEnvironment.STAGING,
      SERVER_PORT: 3456,
      HOST: "https://staging.lindorm.io/",
      PRIO: "ignored",
    };
    developmentConfig = {
      NODE_ENVIRONMENT: NodeEnvironment.DEVELOPMENT,
      SERVER_PORT: 1234,
      HOST: "https://dev.lindorm.io/",
      PRIO: "ignored",
    };
    testConfig = {
      NODE_ENVIRONMENT: NodeEnvironment.TEST,
      SERVER_PORT: 4567,
      HOST: "https://test.lindorm.io/",
      PRIO: "ignored",
    };
    config = new MockConfig({
      environment: environmentConfig,
      production: productionConfig,
      staging: stagingConfig,
      development: developmentConfig,
      test: testConfig,
    });
  });

  test("should provide configuration for production", () => {
    expect(config.get(NodeEnvironment.PRODUCTION)).toMatchSnapshot();
  });

  test("should provide configuration for staging", () => {
    expect(config.get(NodeEnvironment.STAGING)).toMatchSnapshot();
  });

  test("should provide configuration for development", () => {
    expect(config.get(NodeEnvironment.DEVELOPMENT)).toMatchSnapshot();
  });

  test("should provide configuration for test", () => {
    expect(config.get(NodeEnvironment.TEST)).toMatchSnapshot();
  });
});
