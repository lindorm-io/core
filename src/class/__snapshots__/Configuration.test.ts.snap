// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration should provide configuration for development 1`] = `
Object {
  "HOST": "https://dev.lindorm.io/",
  "NODE_ENVIRONMENT": "development",
  "PRIO": "recognized",
  "SERVER_PORT": 1234,
}
`;

exports[`Configuration should provide configuration for production 1`] = `
Object {
  "HOST": "https://lindorm.io/",
  "NODE_ENVIRONMENT": "production",
  "PRIO": "recognized",
  "SERVER_PORT": 2345,
}
`;

exports[`Configuration should provide configuration for staging 1`] = `
Object {
  "HOST": "https://staging.lindorm.io/",
  "NODE_ENVIRONMENT": "staging",
  "PRIO": "recognized",
  "SERVER_PORT": 3456,
}
`;

exports[`Configuration should provide configuration for test 1`] = `
Object {
  "HOST": "https://test.lindorm.io/",
  "NODE_ENVIRONMENT": "test",
  "PRIO": "recognized",
  "SERVER_PORT": 4567,
}
`;
